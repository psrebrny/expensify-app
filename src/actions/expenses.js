import uuid from 'uuid';

export const addExpense = ({description = '', note = '', amount = 0, createdAt = 0}) => {
  
  return {
    type: 'ADD_EXPENSE',
    payload: {
      id: uuid(),
      description,
      note,
      amount,
      createdAt
    }
  };
};

export const removeExpense = (id = '') => {
  return {
    type: 'REMOVE_EXPENSE',
    payload: {id}
  };
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  payload: {id, updates}
});