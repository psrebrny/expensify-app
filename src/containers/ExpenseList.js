import React from 'react';
import { connect } from 'react-redux';
import { ExpenseListItem } from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpenseListClass = (props) => {
  
  const getExpenses = (expenses) => {
    return expenses.map((expense) => {
      return <ExpenseListItem key={expense.id} {...expense}/>;
    });
  };
  
  return (
      <div>
        {
          props.expenses.length === 0 ? (
              <p>No expenses</p>
          ) : (
              getExpenses(props.expenses)
          )
        }
      </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

const ExpenseList = connect(mapStateToProps)(ExpenseListClass);

export { ExpenseList, ExpenseListClass };

