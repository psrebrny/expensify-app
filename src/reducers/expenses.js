// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.payload
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({id}) => id !== action.payload.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.payload.id) {
          return {
            ...expense,
            ...action.payload.updates
          };
        } else {
          return expense;
        }
      });
    case 'GET_EXPENSES':
      return action.payload;
    default:
      return state;
  }
};