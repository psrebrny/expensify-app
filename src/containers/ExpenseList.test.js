import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListClass } from './ExpenseList';
import expenses from '../test/expensesMock';

describe('containers/ExpenseList.js', () => {
  test('render', () => {
    const wrapper = shallow(<ExpenseListClass expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
  });
  
  test('render empty', () => {
    const wrapper = shallow(<ExpenseListClass expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
  });
});