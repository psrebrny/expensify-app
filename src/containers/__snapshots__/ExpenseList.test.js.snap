// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/ExpenseList.js render 1`] = `
<div>
  <Connect(ExpenseListItemClass)
    amount={195}
    createdAt={0}
    description="Gum"
    id={1}
    key="1"
    note=""
  />
  <Connect(ExpenseListItemClass)
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id={2}
    key="2"
    note=""
  />
  <Connect(ExpenseListItemClass)
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id={3}
    key="3"
    note=""
  />
</div>
`;

exports[`containers/ExpenseList.js render empty 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;
